### PR TITLE
Multiport fifo

### PR DESCRIPTION
**Pull request by lekcyjna123**
To have proper implementation of interrupts we are going to need multiport FIFO for FreeRF, so here is a draft of such FIFO.

ToDo:
- [x] Implement MultiportFIFO
- [x] Write one port tests for MultiportFIFO
- [x] Write many port tests for MultiportFIFO
- [x] ~Bonus: create Multi selection module~

### CODE REVIEW

File: coreblocks/utils/fifo.py [hunk 1]
```diff
@@ -108,3 +110,144 @@ def _() -> None:
             m.d.sync += self.write_idx.eq(0)
 
         return m
+
+
+def popcount(m: Module, x: Signal) -> Signal:
```
Comment (lekcyjna123) [inline coreblocks/utils/fifo.py:115]: Move popcount to utils

**Comment by tilk** [comment]
So it's _just an optimization_. ;)

**Comment by lekcyjna123** [comment]
Yes, but I think that this optimization, can simplify implementation, because we will not have to take into account some weird corner cases e.g. RobAllocation want to write on not already dealocated old entry.

**Comment by tilk** [comment]
> because we will not have to take into account some weird corner cases e.g. RobAllocation want to write on not already dealocated old entry.

Now you lost me again.

File: coreblocks/utils/fifo.py [hunk 2]
```diff
@@ -108,3 +122,171 @@ def _() -> None:
...
+        self.depth = depth
+        self.port_count = port_count
+        self.fifo_count = fifo_count
+        self.init = init
+
+        self.connected_reads = 0
+        self.connected_writes = 0
+
+        self._read_methods = [Method(o=self.layout) for _ in range(self.port_count)]
+        self._write_methods = [Method(i=self.layout) for _ in range(self.port_count)]
```
Comment (Arusekk) [inline coreblocks/utils/fifo.py:187]: The methods will be anonymous, maybe we could give them some (partially) useful names?

File: test/utils/test_fifo.py [hunk 3]
```diff
@@ -59,14 +77,14 @@ def source():
...
 
-                v = random.randint(0, (2**fifoc.fifo.width) - 1)
-                yield from fifoc.fifo_write.call(data=v)
-                expq.appendleft(v)
-
                 if random.random() < 0.005:
                     yield from fifoc.fifo_clear.call()
                     expq.clear()
 
+                v = random.randint(0, (2**fifoc.fifo.width) - 1)
```
Comment (Arusekk) [inline test/utils/test_fifo.py:84]: ;-)
```suggestion
                v = random.randrange(2**fifoc.fifo.width)
```

**Comment by lekcyjna123** [comment]
I uploaded tests for multiporting. Additionally there is some weird bug, which cause that input of one of writing ports freeze. This is random error and hard to reproduce, so I will have to dig deeper. In most cases it was connected with sending `clear` signal, but now I have found an reproduction where no clearing was issued.

**Review by Arusekk — COMMENTED** [review]
Better to ask for forgiveness than permission :)

**Comment by lekcyjna123** [comment]
Fixed implementation of multiport fifo.

**Pull request merged**